### PR TITLE
feat(community): add tools support for litellm

### PR DESCRIPTION
**Pull request by igor-drozdov**
I used the following example to validate the behavior

```python
from langchain_core.prompts import ChatPromptTemplate
from langchain_core.runnables import ConfigurableField
from langchain_anthropic import ChatAnthropic
from langchain_community.chat_models import ChatLiteLLM
from langchain_core.tools import tool
from langchain.agents import create_tool_calling_agent, AgentExecutor

@tool
def multiply(x: float, y: float) -> float:
    """Multiply 'x' times 'y'."""
    return x * y

@tool
def exponentiate(x: float, y: float) -> float:
    """Raise 'x' to the 'y'."""
    return x**y

@tool
def add(x: float, y: float) -> float:
    """Add 'x' and 'y'."""
    return x + y

prompt = ChatPromptTemplate.from_messages([
    ("system", "you're a helpful assistant"),
    ("human", "{input}"),
    ("placeholder", "{agent_scratchpad}"),
])

tools = [multiply, exponentiate, add]

llm = ChatAnthropic(model="claude-3-sonnet-20240229", temperature=0)
# llm = ChatLiteLLM(model="claude-3-sonnet-20240229", temperature=0)

agent = create_tool_calling_agent(llm, tools, prompt)
agent_executor = AgentExecutor(agent=agent, tools=tools, verbose=True)

agent_executor.invoke({"input": "what's 3 plus 5 raised to the 2.743. also what's 17.24 - 918.1241", })
```

`ChatAnthropic` version works:

```
> Entering new AgentExecutor chain...

Invoking: `exponentiate` with `{'x': 5, 'y': 2.743}`
responded: [{'text': 'To calculate 3 + 5^2.743, we can use the "exponentiate" and "add" tools:', 'type': 'text', 'index': 0}, {'id': 'toolu_01Gf54DFTkfLMJQX3TXffmxe', 'input': {}, 'name': 'exponentiate', 'type': 'tool_use', 'index': 1, 'partial_json': '{"x": 5, "y": 2.743}'}]

82.65606421491815
Invoking: `add` with `{'x': 3, 'y': 82.65606421491815}`
responded: [{'id': 'toolu_01XUq9S56GT3Yv2N1KmNmmWp', 'input': {}, 'name': 'add', 'type': 'tool_use', 'index': 0, 'partial_json': '{"x": 3, "y": 82.65606421491815}'}]

85.65606421491815
Invoking: `add` with `{'x': 17.24, 'y': -918.1241}`
responded: [{'text': '\n\nSo 3 + 5^2.743 = 85.66\n\nTo calculate 17.24 - 918.1241, we can use:', 'type': 'text', 'index': 0}, {'id': 'toolu_01BkXTwP7ec9JKYtZPy5JKjm', 'input': {}, 'name': 'add', 'type': 'tool_use', 'index': 1, 'partial_json': '{"x": 17.24, "y": -918.1241}'}]

-900.8841[{'text': '\n\nTherefore, 17.24 - 918.1241 = -900.88', 'type': 'text', 'index': 0}]

> Finished chain.
```

While `ChatLiteLLM` version doesn't.

But with the changes in this PR, along with:

- https://github.com/langchain-ai/langchain/pull/23823
- https://github.com/BerriAI/litellm/pull/4554

The result is _almost_ the same:

```
> Entering new AgentExecutor chain...

Invoking: `exponentiate` with `{'x': 5, 'y': 2.743}`
responded: To calculate 3 + 5^2.743, we can use the "exponentiate" and "add" tools:

82.65606421491815
Invoking: `add` with `{'x': 3, 'y': 82.65606421491815}`


85.65606421491815
Invoking: `add` with `{'x': 17.24, 'y': -918.1241}`
responded:

So 3 + 5^2.743 = 85.66

To calculate 17.24 - 918.1241, we can use:

-900.8841

Therefore, 17.24 - 918.1241 = -900.88

> Finished chain.
```

If no one reviews your PR within a few days, please @-mention one of baskaryan, efriis, eyurtsev, ccurme, vbarda, hwchase17.

### CODE REVIEW

**Review by ccurme — COMMENTED** [review]
Hi @igor-drozdov, regarding the integration tests: they perform real requests, but there is no automated process that runs them for `langchain-community` in Langchain's Github. (e.g., we do not maintain credentials for each integration).

My recommendation is to:
- remove `litellm` from dependencies;
- for unit tests (that do run automatically), configure pytest to skip them if `litellm` is not installed (see comment in-line);
- on your end, `pip install litellm` in your env and check that unit + integration tests pass.

Hope this makes sense. Let me know if I can clarify.

File: libs/community/tests/unit_tests/chat_models/test_litellm.py [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+import pytest
+from langchain_core.language_models import BaseChatModel
+from langchain_standard_tests.unit_tests import (
+    ChatModelUnitTests,  # type: ignore[import-not-found]
+)
+
+from langchain_community.chat_models.litellm import ChatLiteLLM
+
+
+class TestLiteLLMStandard(ChatModelUnitTests):
```
Comment (ccurme) [inline libs/community/tests/unit_tests/chat_models/test_litellm.py:14]: Can we configure this to skip if `litellm` is not installed by setting `pytest.mark.requires("litellm")`, similar to how it's done here: https://github.com/langchain-ai/langchain/blob/c4e149d4f18319bef6f1d4c409250c4a0ad21dac/libs/community/tests/unit_tests/chat_models/test_anthropic.py#L17

**Comment by igor-drozdov** [comment]
@ccurme ah, I've noticed that `extended_testing_deps.txt` should be updated as well, I've added `litellm` there, but still got the import problem because `import litellm` raises:

```
ImportError: cannot import name 'model_validator' from 'pydantic'
```

It happened because pydantic v1 is used, but when I try updating to v2, there's a conflict with this library: https://pypi.org/project/javelin-sdk/0.2.5/#description

I decreased the version to the one that doesn't raise the import error and now the tests pass.

Thank you for your reviews! Could you please have another look?

File: libs/community/langchain_community/chat_models/litellm.py [hunk 2]
```diff
@@ -132,10 +136,30 @@ def _convert_delta_to_message_chunk(
     else:
         additional_kwargs = {}
 
+    tool_call_chunks = []
+    if raw_tool_calls := _dict.get("tool_calls"):
+        additional_kwargs["tool_calls"] = raw_tool_calls
```
Comment (ccurme) [inline libs/community/langchain_community/chat_models/litellm.py:141]: `raw_tool_calls` should probably be a dict or something readily json serializable. I think we may need to call `.model_dump()` on the tool calls.

**Review by ccurme — COMMENTED** [review]
Looks good, one question on `raw_tool_calls`.

I noticed that tool_calls aren't implemented outside of streaming (e.g., within `_convert_dict_to_message`). Can do in a separate PR though if needed.

File: libs/community/langchain_community/chat_models/litellm.py [hunk 3]
```diff
@@ -132,10 +136,30 @@ def _convert_delta_to_message_chunk(
...
+            tool_call_chunks = [
+                ToolCallChunk(
+                    name=rtc.function.name,
+                    args=rtc.function.arguments,
+                    id=rtc.id,
+                    index=rtc.index,
+                )
+                for rtc in raw_tool_calls
+            ]
+        except KeyError:
```
[the remark this answers is not in the document]
Comment (igor-drozdov) [reply libs/community/langchain_community/chat_models/litellm.py:152]: Now, when `rtc["function"]` calls are in place, this exception handling is legit

**Review by ccurme — APPROVED** [review]

**Pull request merged**